### ISSUE TITLE
Add version suffix to last.ckpt

### ISSUE THREAD

**Issue by carmocca** [PyTorchLightning/pytorch-lightning#5030]
## 🚀 Feature

If you use `ModelCheckpoint(save_last=True)` and you run an experiment twice in the same directory, then this set of checkpoints is generated:

```
file-v0.ckpt
file-v1.ckpt
...
last.ckpt (the last of the second run)
```

the idea is to add a version also to `last.ckpt` if it would get overwritten:

```
file-v0.ckpt
file-v1.ckpt
...
last-v0.ckpt (last of the first run)
last-v1.ckpt (last of the second run)
```

### Motivation

Avoid overwriting existing checkpoints

### Alternatives

Modifying the default `CHECKPOINT_NAME_LAST` to avoid the conflict

### Additional context

Discussed in https://github.com/PyTorchLightning/pytorch-lightning/issues/5000#issuecomment-740701415

To be done after #5008 is merged.

cc @borda @carmocca @awaelchli @ninginthecloud @jjenniferdai @rohitgr7

**Comment by carmocca** [comment]
@awaelchli answering https://github.com/PyTorchLightning/pytorch-lightning/issues/5000#issuecomment-741503135 here

How do you propose we fix this then?

**Comment by awaelchli** [comment]
@carmocca Before I make a suggestion could you help me understand the context a bit more. Is this issue a) about damage control for users that accidentally write to the same directory, or b) is this about a particular use case where a user is doing this on purpose? 

I am asking this because of two reasons. 1) I notice in your description you didn't append epoch numbers to the filenames and the primary reason why the save_last option is exists is for sequential checkpoints numbered by some metric like epoch or step. 2) last.ckpt should always be a copy of the last epoch, i.e. epoch_N.ckpt is identical to last.ckpt. Therefore it would be save to overwrite last.ckpt in a subsequent run.

Note that situation a) cannot be solved by appending version affixes to the files. Consider the case where user changes epoch size between runs so they would get a mix of files and you wouldn't be able to tell if epoch=x_step=y.ckpt or epoch=x_step=z.ckpt belonged to the first or second run.

**Comment by awaelchli** [comment]
One may consider this option:
```
file-v0.ckpt
file-v1.ckpt
...
last-v0.ckpt (last of the first run)
last-v1.ckpt (last of the second run)
last.ckpt (copy of last-v1)
```

**Comment by carmocca** [comment]
It is about a)

>  I notice in your description you didn't append epoch numbers

Sorry, I didn't for brevity

> 2) last.ckpt is always a copy of the last epoch, i.e. epoch_N.ckpt is identical to last.ckpt. Therefore it would be save to overwrite last.ckpt in a subsequent run.

That is not the case if you are monitoring something. For example, if you run an experiment for 10 epochs and your best performance is `epoch_4.ckpt`, `last.ckpt` corresponds to the tenth epoch. We discussed this in https://github.com/PyTorchLightning/pytorch-lightning/issues/4335#issuecomment-731884628

> Consider the case where user changes epoch size between runs so they would get a mix of files and you wouldn't be able to tell if epoch=x_step=y.ckpt or epoch=x_step=z.ckpt belonged to the first or second run. One would only be able to tell by inspecting the individual checkpoints for hyperparameters.

This is impossible already in master, version suffixes do not imply that all `-v0` files necessarily correspond to the first run, and `-v1` to the second run, etc. It is just a means of avoiding overwritting files.

**Comment by carmocca** [comment]
> One may consider this option:
> 
> ```
> file-v0.ckpt
> file-v1.ckpt
> ...
> last-v0.ckpt (last of the first run)
> last-v1.ckpt (last of the second run)
> last.ckpt (copy of last-v1)
> ```

if `last.ckpt` must stay, then I would keep it as is and make version suffixes start at 1.

However, this would be inconsistent with what was discussed in #5008.

So should we just start versions at 1 and forget about renaming files?
cc @rohitgr7 @Borda

**Comment by awaelchli** [comment]
> For example, if you run an experiment for 10 epochs and your best performance is epoch_4.ckpt, last.ckpt corresponds to the tenth epoch. 

This is correct behaviour. last.ckpt as the name suggests should point to the last saved checkpoint in a run, in this case it would be epoch 10 (epoch=9.ckpt). Again, last.ckpt is there for convenience if the user expects their run to be interrupted (for example manually) they can easily restore and continue training with last.ckpt without having to look up the exact filename. Having a file best.ckpt is a separate feature one could consider.

**Comment by carmocca** [comment]
Exactly, but you might not have saved on your tenth epoch because your model wasnt in the top-k

> last.ckpt is always a copy of the last epoch, i.e. epoch_N.ckpt is identical to last.ckpt. Therefore it would be save to overwrite last.ckpt in a subsequent run.

So `epoch_N.ckpt` doesnt exists but `last.ckpt` does. Then it is not safe to overwrite `last.ckpt`

**Comment by awaelchli** [comment]
These edge cases are terrible. Looks like the versioning is necessary then.

**Comment by rohitgr7** [comment]
No stong opinion here. I am fine with both https://github.com/PyTorchLightning/pytorch-lightning/issues/5030#issuecomment-741763874 renaming or not, 0 or 1 as the first prefix version.

**Comment by edenlightning** [comment]
@carmocca is this done?

**Comment by carmocca** [comment]
> @carmocca is this done?

No.
